File: backend/schemas/exam.py
from typing import Any
from pydantic import BaseModel


class GenerateExamRequest(BaseModel):
    questionCount: int = 75

    timed: bool = False

    durationMinutes: int | None = None


class GenerateExamResponse(BaseModel):
    success: bool

    examId: str

    questionCount: int | None = None

    unfinishedExam: bool = False


class ExamSession(BaseModel):
    examId: str
    questionCount: int

    questions: list[dict[str, Any]]

    answers: dict[str, Any]

    completed: bool

    startedAt: str

    timed: bool

    durationMinutes: int | None

    remainingSeconds: int | None = None


class SaveAnswerRequest(BaseModel):
    questionId: str
    selectedOption: str | None

class MarkReviewRequest(BaseModel):
    questionId: str
    marked: bool

class SubmitExamResponse(BaseModel):
    success: bool
    score: int
    totalQuestions: int
    correctAnswers: int
    wrongAnswers: int
    unanswered: int
    percentage: float

File: frontend/src/types/exam.ts
export interface ExamQuestion {
  id: string;
  number: number;
  page?: number;
  subject?: string;
  question: string;
  options: Record<string, string>;
}

export interface AnswerState {
  selectedOption: string | null;
  visited: boolean;
  markedForReview: boolean;
}

export interface ExamSession {
  examId: string;
  questionCount: number;

  questions: ExamQuestion[];

  answers: Record<string, AnswerState>;

  completed: boolean;

  startedAt: string;

  timed: boolean;

  durationMinutes: number | null;
}

export interface GenerateExamRequest {
  questionCount: number;
  timed: boolean;
  durationMinutes: number | null;
}

export interface GenerateExamResponse {
  success: boolean;
  examId: string;
  questionCount: number;
}

export interface SaveAnswerRequest {
  questionId: string;
  selectedOption: string | null;
}

export interface MarkReviewRequest {
  questionId: string;
  marked: boolean;
}

export interface SubmitExamResponse {
  success: boolean;
  score: number;
  totalQuestions: number;
  correctAnswers: number;
  wrongAnswers: number;
  unanswered: number;
  percentage: number;
}